### ISSUE TITLE
GSoC Goals / Timelines / Milesstones

### ISSUE THREAD

**Issue by eddelbuettel** [Yashwants19/RcppMLPACK#2]
@Yashwants19 @coatless @rcurtin  :  shall we hash something out and out it here now that we're official?

**Comment by rcurtin** [comment]
Agreed, I think that would be a good idea.  I imagine that this repository in particular will end up being unimportant for the actual project, since that will all be done in the main repository.  Maybe we should use @Yashwants19's fork at https://github.com/Yashwants19/mlpack?

One giant PR (like https://github.com/mlpack/mlpack/pull/1884 and https://github.com/mlpack/mlpack/pull/1949) is probably super unwieldy, so maybe it makes sense to set out a bunch of goals as issues, and then make PRs one-by-one to merge into a branch like `Yashwants19/r-bindings`, and then when everything is ready, we just open a big PR for `Yashwants19/r-bindings` to the main mlpack repository, give it a once-over, and merge it in?

@Yashwants19 you're driving this effort, so, up to you how you want to do it, but in any case it would be great to have some way that all four of us can easily collaborate, review code, and discuss. :)

**Comment by eddelbuettel** [comment]
I didn't even think of that, but yes.  As the bindings part moves and changes we can equally well work over there.  And it is of course the "home project" for the summer.
 
On the margin smaller codebases compiler faster :)  But whatever works.

**Comment by rcurtin** [comment]
> On the margin smaller codebases compiler faster :) But whatever works.

I'd agree with that, but in this case mlpack's automatic binding generator is integrated with the library, and we haven't put any thought into how to modularize it / move it to its own repository.  Maybe someday?

**Comment by coatless** [comment]
The timeline and goals was detailed on page 11 of the final proposal. 

![Screen Shot 2020-05-26 at 5 40 50 PM](https://user-images.githubusercontent.com/833642/82957196-12e34e80-9f78-11ea-929a-77490c3230a2.png)

I think @Yashwants19 is expecting to be able to fully focus on code starting June 1st? That said, the PoC `RcppMLPACK` package covers part of the first coding period deliverables.

**Comment by Yashwants19** [comment]
Hi I am really sorry for late response. I was messed up with [go-bindings](https://github.com/mlpack/mlpack/pull/1884) work, but it's almost complete and ready to merge. 
BTW @rcurtin ideas seems perfect to me, I will create a branch and an issue with all the goals, and then I will create PRs one-by-one that can be reviewed and merge into the respective branch. But for completing R utility functions, I think so we must do this here (Yashwants19/RcppMLPACK) only because it will be easy to test them here in proper environment. After this I can switch all the work(all cmake process, glue-code, documentation and testing with azure ) to the branch.

If you all agree with the timeline/goals given in proposal we can follow the same, but if any one want to change some part I am totally fine with that.

Thank You.

**Comment by eddelbuettel** [comment]
Sounds like a reasonable plan as you need to work on both sides.  

Maybe we can use issue tickets here for (simple) progress reports / updates?

**Comment by Yashwants19** [comment]
Milestone/ Goals / Timelines :
First Evaluation Period:

1) WEEK --> 1
   OPEN a PR in RcppMLPACK.
   - [x]  Completing R utility functions and testing of these functions in RcppMLPACK.
      -- This week my main focus will to pass matrices from/to R without creating unnecessary copies.
   - [x]  Some more hand written algorithm (with all r_util functions like [Matrix_with_info](https://github.com/mlpack/mlpack/blob/1e802b27b28e555f7a70ac450bf2cf3b63870b64/src/mlpack/core/util/param.hpp#L914))
      -- Discuss with mentors whether the auto-generated .cpp and .R must look like hand written binding or some more improvement is still required.
   
2) WEEK --> 2 - 3 
   After completing the above work I will open PRs one-by-one in a branch as suggested by @rcurtin.
   - [x]   Complete tests (if some tests still lefts)
   - [x]   Prepare complete directory structure using cmake in mlpack :
      -- Replace version in DESCRIPTION using cmake (${MLPACK_VERSION}).
      -- Build complete NAMESPACE file using cmake only.
      -- Copying all the mlpack files to RcppMLPACK/src/mlpack
      -- Generate RcppExports.cpp/.R if possible didn't explore this part yet.
      -- Build config file from config.ac
    - [x]  After the generating the directory, Install/Build the package using R CMD.
    - [x]  Test the package with handwritten bindings (Initially).
 
 3) WEEK --> 4
    - [x]  Write the glue code for generating automating-bindings.
       -- print_cpp.cpp
       -- print_R.cpp
       -- All the supportive files.
    - [x]  Integrate these glue code with cmake to build generating automating-bindings using:
       -- generate_cpp.cpp.in
       -- generate_R.cpp.in
    I will follow the same approach that is followed by other [bindings](https://github.com/mlpack/mlpack/tree/master/src/mlpack/bindings).
 

These milestones is up to first evaluations period. I will update this as soon as I complete the work/ first evaluations completes.

Thank You.

**Comment by eddelbuettel** [comment]
Looks good!  Nice list overall.  Should we break smaller points discussion off somewhere or muddle it all here?   (For week one: "pass matrices from/to R without creating unnecessary copies".  That already happens that way via RcppArmadillo using the advanced constructor taking the R memory of a (suitable: int or double) vector.  I can expand on that if needed.

**Comment by Yashwants19** [comment]
Hi, @eddelbuettel  as you suggest -- For discussion I can open issues one-by-one for every week or we can discussion on the respective PR ?

I figured out some approaches:
For avoiding copy from R to C++
1) Tow Cast: SEXP --> NumericMatrix --> armadillo(Advance constructor) 
   (In microbenchmark it was looking best, but with a minor difference only).
2) One Cast: NumericMatrix(Reference) --> armadillo(Advance constructor)
3) Direct reference of armadillo matrix.
Please correct me If I am doing something wrong here.
Which approach should I follow ?


But for avoiding copy from C++ to R, I want to ask some questions.
1) When we pass reference of arma to R like I did [here](https://github.com/Yashwants19/RcppMLPACK/blob/a1c666d33bc6679670ed9a4a0cbac99a33af0ed8/src/r_util.cpp#L84), do this will avoid copy?
2) Do this method will handle preallocated memory (If it use preallocated memory)?
    Or we have to take care of this, like mlpack did for  other bindings?
3) If we convert arma to NumericMatrix is this will be helpful?

Example :
```
NumericMatrix/SEXP CLI_GetParamMat(const std::string& paramName)
{
  arma::mat& m = CLI::GetParam<arma::mat>(paramName);
  if (m.n_elem <= arma::arma_config::mat_prealloc)
  {
    double* newMem = new double[m.n_elem];
    arma::arrayops::copy(newMem, m.mem, m.n_elem);
    NumericMatrix mat(m.n_rows, m.n_cols, newMem);
    return mat / (SEXP)mat;
  }
  else
  {
    arma::access::rw(m.mem_state) = 1;
    NumericMatrix mat(m.n_rows, m.n_cols, m.memptr());
    return mat / (SEXP)mat;
  }
}
``` 

Thank You.

**Comment by eddelbuettel** [comment]
I am also happy to just pile it up here but a single thread risks being long. 

In brief, and I though we had a benchmark file in the repo but apparently not -- will look from some tutorials/workship you can go directly -- in this the matrix should not get copied:

```r
R> Rcpp::cppFunction("double topleft(const arma::mat & M) \   #linebreak for display
    { return M(0,0); }", depends="RcppArmadillo")
R> M1 <- matrix(1,2,2)
R> M2 <- matrix(100,10,10)
R> M3 <- matrix(10000,100,100)
R> library(rbenchmark)
R> benchmark(m1=topleft(M1), m2=topleft(M2), m3=topleft(M3), replications=1e6)[,1:4]
  test replications elapsed relative
1   m1      1000000   2.402    1.000
2   m2      1000000   2.410    1.003
3   m3      1000000   2.406    1.002
R> 
```

You can expand the benchmark but that should be both the most performant -- and shortest :) -- form.

**Comment by eddelbuettel** [comment]
Might be better to have one issue or thread per question.  As for the return, I also tend to write code as you did on L84 (the link you show above).  _Purest_ performance is better passing by reference rather than return via the heap but I don't find it as clear. ("Eventually" move semantics and copy elision will help, but not with C++11).

**Comment by coatless** [comment]
@Yashwants19 if it's okay, I'm going to split these off into an issue ticket per week based on https://github.com/Yashwants19/RcppMLPACK/issues/2#issuecomment-636495013 for easier tracking.

**Comment by coatless** [comment]
On the memory path, from _R_ to _C++_, it's easier to directly reference the matrix with the standard `arma::mat&` descriptor in the function definition. We've sorta hit this point from many angles over the years. ([1](https://stackoverflow.com/questions/50170166/deciding-between-numericvector-and-armavec-in-rcpp/50181097#50181097), [2](https://stackoverflow.com/questions/24112893/within-c-functions-how-are-rcpp-objects-passed-to-other-functions-by-referen/24118024#24118024), [3](https://stackoverflow.com/questions/48363076/declare-a-variable-as-a-reference-in-rcpp/48363220#48363220) lay the framework for the memory model)

From  _C++_ to _R_, I'm not sure about having a return type of `arma::mat&`. I'm going to slightly deviate from the L84 remark by @eddelbuettel. That said, I'd trust him more than me on how well supported that return is. 

----

I've avoided `arma::mat&` return approach with _R_ due to a slight concern that once the code block is done executing the reference will go out of scope and point somewhere we're not desiring. Outside of this fear, I'm not sure the object's memory location gets preserved from this advanced CTOR trick back into _R_. 

As an example, consider the following code snippet that I think matches well with the scenario;

```cpp
#include <RcppArmadillo.h>

// [[Rcpp::depends(RcppArmadillo)]]

// [[Rcpp::export]]
arma::mat& memory_reference_mat_return(arma::mat& x) {
  arma::mat& m = x;
  
  return m;
}
```

The function receives a reference from _R_ and attempts to send back the same reference with `arma::mat& m`. 

To assist, let's look at the function call under data with a memory trace activated.

```r
M4 = matrix(1000000,1000,1000)

tracemem(M4)
# <0x7f882c250000>

arma_M4 = memory_reference_mat_return(M4)

tracemem(arma_M4)
# <0x7f882c9f2000>
```

Note, the memory address changed on the reference when it was returned under the pass.

Poking at it a bit more using `.Internal()`, the allocations differ a bit: 

```r
.Internal(inspect(M4))

# @7f882c250000 14 REALSXP g1c7 [MARK,REF(203),TR,ATT] (len=1000000, tl=0) 1e+06,1e+06,1e+06,1e+06,1e+06,...
# ATTRIB:
#  @7f882a1e2678 02 LISTSXP g1c0 [MARK,REF(1)] 
#    TAG: @7f883c00e480 01 SYMSXP g1c0 [MARK,REF(11327),LCK,gp=0x4000] "dim" (has value)
#    @7f8841e34840 13 INTSXP g1c1 [MARK,REF(65535)] (len=2, tl=0) 1000,1000

.Internal(inspect(arma_M4))
# @7f882c9f2000 14 REALSXP g1c7 [MARK,REF(4),TR,ATT] (len=1000000, tl=0) 1e+06,1e+06,1e+06,1e+06,1e+06,...
ATTRIB:
#   @7f882a1d7558 02 LISTSXP g1c0 [MARK,REF(1)] 
#     TAG: @7f883c00e480 01 SYMSXP g1c0 [MARK,REF(11327),LCK,gp=0x4000] "dim" (has value)
#     @7f8841e36638 13 INTSXP g1c1 [MARK,REF(65535)] (len=2, tl=0) 1000,1000
```

The above can be broken down to:

```
Address          Type Name    Extra                                True Length
|                |            |                                    |
+-----------+    +----+       +--------------------+               +--+
@7f882c250000 14 REALSXP g1c7 [MARK,REF(203),TR,ATT] (len=1000000, tl=0) 1e+06,1e+06,...
              ++         +--+                        +-----------+       +------------+
              |          |                            |                  |
              Type       GC Info                     Length              Data
```

Diagram modified from https://www.brodieg.com/2019/02/18/an-unofficial-reference-for-internal-inspect/

I think maybe to avoid the `arma::mat&` return and use `NumericMatrix`.

**Comment by eddelbuettel** [comment]
@coatless: And ... so what?  Doesn't that mostly show that the few bytes of the SEXP container change but what we care about is the (potentially much larger) matrix blob referenced therein---and that reference just gets copied -- which is lightweight and what we want.   Maybe I am missing something...

**Comment by coatless** [comment]
@eddelbuettel that is indeed exactly what it shows. Digging into the source for RcppArmadillo, does yield the proper convert:

https://github.com/RcppCore/RcppArmadillo/blob/7aaba188e678cb173390c12a91c734065d20e0a3/inst/include/RcppArmadilloWrap.h#L30-L54

And a better example sin the mem trace that led into the bowels of _R_ would be: 

```cpp
#include <RcppArmadillo.h>
// [[Rcpp::depends(RcppArmadillo)]]

// [[Rcpp::export]]
arma::mat& ref_return(arma::mat& x) {
  return x;
}

// [[Rcpp::export]]
arma::mat copy_return(arma::mat& x) {
  return x;
}
```

```r
M4 = matrix(runif(1e8),1e4,1e4)

micro_timings = microbenchmark::microbenchmark(
  ref = ref_return(M4),
  copy = copy_return(M4)
  )

summary(micro_timings)
ggplot2::autoplot(micro_timings)
```

|expr   |      min|       lq|     mean|    median|        uq|       max| neval|cld |
|:------|--------:|--------:|--------:|---------:|---------:|---------:|-----:|:---|
|ref    | 383.0586| 425.3945|  499.494|  513.9527|  546.9611|  777.4962|   100|a   |
|copy | 884.2503| 935.8164| 1035.057| 1035.9380| 1081.6492| 1676.8585|   100|b   |

![image](https://user-images.githubusercontent.com/833642/83453136-93de9200-a41f-11ea-94d2-93717940f4e7.png)

So, long story short @Yashwants19, please make direct use of `arma::mat&` for both parameter and return definitions!

**Comment by eddelbuettel** [comment]
How does that gel with your earlier recommendation:

> I think maybe to avoid the arma::mat& return and use NumericMatrix.

In your followup those weren't even the options compared.  But regardless we all seem to agree that using references is best.  Hooray.

**Comment by coatless** [comment]
@eddelbuettel in short: Use the built-in `wrap()` converter function associated with `arma::mat&` and avoid writing custom logic for `NumericMatrix`.

---

In long...

Based on [`RcppArmadillo`'s `wrap()` code](https://github.com/RcppCore/RcppArmadillo/blob/7aaba188e678cb173390c12a91c734065d20e0a3/inst/include/RcppArmadilloWrap.h#L30-L54), going the `NumericMatrix` route would have added additional class overhead compared to `RcppArmadillo`'s direct use of the base  `RObject` policy. Comparing the approaches would yield...

```cpp
#include <RcppArmadillo.h>
// [[Rcpp::depends(RcppArmadillo)]]

// [[Rcpp::export]]
arma::mat& armadillo_ref_return(arma::mat& x) {
  return x;
}

// [[Rcpp::export]]
Rcpp::NumericMatrix rcpp_ref_return(arma::mat& x) {
  
  // Not the same memory reuse trick Armadillo uses.
  // Though, this would be the closest available with Rcpp Matrix class ctor
  Rcpp::NumericMatrix out( x.n_rows , x.n_cols, x.begin() );

  return out;
}
```

```r
M4 = matrix(runif(1e8),1e4,1e4)

micro_timings = microbenchmark::microbenchmark(
  armadillo_ref = armadillo_ref_return(M4),
  numericmatrix_ref = rcpp_ref_return(M4)
  )
```

|expr              |      min|       lq|     mean|   median|       uq|      max| neval|cld |
|:-----------------|--------:|--------:|--------:|--------:|--------:|--------:|-----:|:---|
|armadillo_ref     | 387.6657| 434.5813| 510.8630| 534.0819| 561.8819| 890.8463|   100|a   |
|numericmatrix_ref | 418.6763| 458.7645| 534.0964| 536.9237| 583.4055| 880.5332|   100|a   |

![image](https://user-images.githubusercontent.com/833642/83562615-81c42880-a4df-11ea-81ee-c2c29bc5af75.png)




That said, if the object isn't a reference in the first place, don't mess with references. In particular, we have:

```cpp
#include <RcppArmadillo.h>
// [[Rcpp::depends(RcppArmadillo)]]

// [[Rcpp::export]]
arma::mat& bad_ref_return(arma::mat x) {
  return x;
}
```

The compiler would trigger: 

> warning: reference to stack memory associated with parameter 'x' returned [-Wreturn-stack-address]

Such a warning indicates when the code block is done executing the reference goes out of scope and will point to a memory location that isn't desired.

**Comment by eddelbuettel** [comment]
> In short: Use the built-in `wrap` (fixed) converter associated with `arma::mat&` and avoid writing custom logic for `NumericMatrix`.

Yes. And I think that was recommendation all along. I am confused why/how three full screens full of mostly R internals most likely entirely foreign to our dear student help anyone.  Maybe best to stop this now.  Esoteric topic for esoteric places.  This is a student internship.

**Comment by coatless** [comment]
> Maybe best to stop this now. Esoteric topic for esoteric places.

Agreed. @Yashwants19 can we help or clarify anything else?

**Comment by Yashwants19** [comment]
I am really really sorry for late response. From last few days my internet is not working that great. 
Thank you for the great explanation. I will follow this approach as suggested. 
Thank you @coatless for opening issue tickets. From now onward If I want to ask any question I will use those tickets only. I almost completed my work of this week, I will open a PR soon may be today or tomorrow.
Again I am really sorry for late response. 
And once again thanks a lot for quick response and this awesome explanation.

Thank you.

**Comment by rcurtin** [comment]
> Yes. And I think that was recommendation all along. I am confused why/how three full screens full of mostly R internals most likely entirely foreign to our dear student help anyone. Maybe best to stop this now. Esoteric topic for esoteric places. This is a student internship.

I learned a lot :smile: (but I do agree, seems like passing references back and forth should be just fine---in the end the big deal is just that we avoid copying the actual memory blob, which it seems like either approach manages to do)

**Comment by rcurtin** [comment]
Also, it seems like we've split out the remaining issues here into #3, #4, and #5---should we close this issue then?  (If it's better to leave it open, no problem from my end, I just like issue gardening :))

**Comment by coatless** [comment]
Yes, we should close out old issues. @Yashwants19 will have to do it as we do not have collaborator status on the repository.

**Comment by eddelbuettel** [comment]
Also re the reference passing back and forth.   `std::move()` may help.  I am slowly reading bits and pieces of Stroustrup's "A Tour of C++", 2nd ed, which is C++14-aware and that seems to be the recommendation.  Haven't had time to play much or benchmark but _maybe_ some of our old rules of thumb can be retired one day...
